music_assistant/server/providers/soundcloud/__init__.py

```diff
@@ -0,0 +1,338 @@
+            self.logger.debug(error)
+        return playlist
+
+    async def get_playlist_tracks(self, prov_playlist_id) -> list[Track]:
+        """Get all playlist tracks for given playlist id."""
+        playlist_obj = await self._soundcloud.get_playlist_details(playlist_id=prov_playlist_id)
+        if "tracks" not in playlist_obj:
+            return []
+        tracks = []
+        for index, item in enumerate(playlist_obj["tracks"]):
+            song = await self._soundcloud.get_track_details(item["id"])
+            try:
+                track = await self._parse_track(song[0])
+                if track:
+                    track.position = index
+                    tracks.append(track)
+            except (KeyError, TypeError, InvalidDataError, IndexError) as error:
+                self.logger.debug("get_playlist_tracks")
+                self.logger.debug(track)
+                self.logger.debug(error)
+                continue
+        return tracks
+
+    async def get_artist_toptracks(self, prov_artist_id) -> list[Track]:
+        """Get a list of 25 most popular tracks for the given artist."""
+        tracks_obj = await self._soundcloud.get_popular_tracks_user(
+            user_id=prov_artist_id, limit=25
+        )
+        tracks = []
+        for item in tracks_obj["collection"]:
+            song = await self._soundcloud.get_track_details(item["id"])
+            try:
+                track = await self._parse_track(song[0])
+                tracks.append(track)
+            except (KeyError, TypeError, InvalidDataError, IndexError) as error:
+                self.logger.debug("get_artist_toptracks")
+                self.logger.debug(track)
+                self.logger.debug(error)
+                continue
+        return tracks
+
+    async def get_similar_tracks(self, prov_track_id, limit=25) -> list[Track]:
+        """Retrieve a dynamic list of tracks based on the provided item."""
+        tracks_obj = await self._soundcloud.get_recommended(track_id=prov_track_id, limit=limit)
+        tracks = []
+        for item in tracks_obj["collection"]:
+            song = await self._soundcloud.get_track_details(item["id"])
+            try:
+                track = await self._parse_track(song[0])
+                tracks.append(track)
+            except (KeyError, TypeError, InvalidDataError, IndexError) as error:
+                self.logger.debug("get_similar_tracks")
+                self.logger.debug(track)
+                self.logger.debug(error)
+                continue
+
+        return tracks
+
+    async def get_stream_details(self, item_id: str) -> StreamDetails:
+        """Return the content details for the given track when it will be streamed."""
+        track_details = await self._soundcloud.get_track_details(track_id=item_id)
+        stream_format = track_details[0]["media"]["transcodings"][0]["format"]["mime_type"]
+        url = await self._soundcloud.get_stream_url(track_id=item_id)
```

Comment:
is this stream url not exportable ? if so set the correct expiration. I believe the default is 1 hour

Reply:
> is this stream url not expirable ?

I don't know...

Where do I set this?

Reply:
leave it like this. if any issues with streaming are reported we'll look into this.